### PR TITLE
Merge of feature/hard-fork-2 to dev branch

### PR DESCRIPTION
**Pull request by tgrospic**
## Overview

This PR is combines all changes on `feature/hard-fork-2` branch to the main `dev` branch.

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by zsluedem** [comment]
@tgrospic there are conflicts now. Do you want me to resolve and you focus on finalization?

**Comment by tgrospic** [comment]
> @tgrospic there are conflicts now. Do you want me to resolve and you focus on finalization?

I'm already in the process of resolving conflicts.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala [hunk 1]
```diff
@@ -261,6 +261,7 @@ object SystemProcesses {
 
           produce(Seq(errorMessage), ack)
 
+        // TODO: Invalid type for address should throw error!
```
Comment (tgrospic) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala:264]: This should be fixed for HF2.

File: rspace/src/main/scala/coop/rchain/rspace/hashing/StableHashProvider.scala [hunk 2]
```diff
@@ -11,48 +15,68 @@ object StableHashProvider {
...
+    //      )
+    //    )
+    val encodedSeq = Seq(
+      channel,
+      serializeA.encode(datum),
+      bool(8).encode(persist).map(_.bytes).getUnsafe
     )
+//    Blake2b256Hash.create(encodedSeq)
+    val encoded = codecSeqByteVector.encode(encodedSeq).getUnsafe.toByteVector
+    Blake2b256Hash.create(encoded)
```
Comment (tgrospic) [inline rspace/src/main/scala/coop/rchain/rspace/hashing/StableHashProvider.scala:80]: [TG] Review hashing changes.

File: rspace/src/main/scala/coop/rchain/rspace/trace/Event.scala [hunk 3]
```diff
@@ -112,10 +112,13 @@ object Consume {
...
-    val channelsByteVectors: Seq[ByteVector] = toOrderedByteVectors(channels)
+//    val channelsByteVectors: Seq[ByteVector] = toOrderedByteVectors(channels)
+    // Hash sorted serialized channels and sort by serialized hash
+    val channelsHashes        = hashSeq(channels)
+    val channelsEncodedSorted = channelsHashes.map(_.bytes)
     new Consume(
-      channelsByteVectors.map(Blake2b256Hash.create),
-      hash(channelsByteVectors, patterns, continuation, persistent),
+      channelsHashes,
+      hash(channelsEncodedSorted, patterns, continuation, persistent),
```
Comment (tgrospic) [inline rspace/src/main/scala/coop/rchain/rspace/trace/Event.scala:121]: [TG] Review channel hashes.

File: node/src/test/scala/coop/rchain/node/mergeablity/MergingBranchMergerSpec.scala [hunk 4]
```diff
@@ -90,326 +135,353 @@ class MergingBranchMergerSpec extends FlatSpec with Matchers {
...
+  //         |*
+  //         | """.stripMargin
+  //
+  //    val layers = graph.split('\n')
+  //    val template = layers
+  //      .map(_.toList.zipWithIndex.filter((c) => c._1 == '*').map(_._2))
+  //      .filter(_.nonEmpty)
+  //      .reverse
+  //    getRejectedBlocksAtTheTop(template)
+  //  }
```
Comment (tgrospic) [inline node/src/test/scala/coop/rchain/node/mergeablity/MergingBranchMergerSpec.scala:395]: Review commented tests in MergingBranchMergerSpec.

File: casper/src/main/resources/RevVault.rho [hunk 5]
```diff
@@ -253,17 +253,17 @@ in {
...
           // Deposits the entire balance in the fromPurse to the toPurse.
-          contract _depositTemplate(toPurse, fromPurse, retCh) = {
+          contract _depositTemplate(fromPurse, toPurse, retCh) = {
             new amountCh, depositSuccessCh in {
-              fromPurse!("getBalance", *amountCh) |
+              toPurse!("getBalance", *amountCh) |
               for (@amount <- amountCh) {
                 // Calls purse deposit method in MakeMint contract; adds amount to toPurse and subtracts from fromPurse.
-                toPurse!("deposit", amount, *fromPurse, *depositSuccessCh) |
+                fromPurse!("deposit", amount, *toPurse, *depositSuccessCh) |
```
Comment (tgrospic) [inline casper/src/main/resources/RevVault.rho:266]: Purses from/to renaming is wrong.

**Review by tgrospic — COMMENTED** [review]
All changes look good for single parent. I've added a few minor comments that can be addressed later.

**Review by nzpr — APPROVED** [review]
LGTM

**Comment by zsluedem** [comment]
LGTM

**Review by zsluedem — APPROVED** [review]

**Comment by tgrospic** [comment]
bors merge

**Comment by zsluedem** [comment]
:100:

**Pull request merged**